Fix zdb/ztest -o and refactor libzpool::set_global_var

problame opened the pull request:
```
commit ef7f57852233dd603d022e83505f9791fe2b107b (HEAD -> zil-pmem-tmp, problame/fix-zdb--o)
Author: Christian Schwarz <me@cschwarz.com>
Date:   Tue Feb 16 11:14:44 2021 +0100

    ztest: fix -o by calling set_global_var in child processes
    
    Without set_global_var() in the child processes the -o option provides
    little use.
    
    Before this change set_global_var() was called as a side-effect of
    getopt processing which only happens for the parent ztest process.
    
    This change limits the set of options that can be set and makes them
    available to the child through ztest_shared_opts_t.
    
    Future work: support arbitrary option count and length.
    
    Signed-off-by: Christian Schwarz <me@cschwarz.com>

commit 0c2eb4c086fd3d0eab7cd6356516c226b3d12ce6
Author: Christian Schwarz <me@cschwarz.com>
Date:   Tue Feb 16 12:27:48 2021 +0100

    libzpool: set_global_var: refactor to not modify 'arg'
    
    Also fixes leak of the dlopen handle in the error case.
    
    Signed-off-by: Christian Schwarz <me@cschwarz.com>

commit 0b750aaf0564b9be87f3ee4a0ac07cabf951f24f
Author: Christian Schwarz <me@cschwarz.com>
Date:   Mon Feb 15 13:02:32 2021 +0100

    libzpool: set_global_var: fix endianness handling (fixes zdb -o )
    
    Without this patch I get the error
    
      Setting global variables is only supported on little-endian systems
    
    when using `zdb -o` on my amd64 machine.
    
    Signed-off-by: Christian Schwarz <me@cschwarz.com>

```

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactoring

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- ~~I have updated the documentation accordingly.~~
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
-  ~~I have run the ZFS Test Suite with this change applied.~~
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### Notes To Review

I actually don't know what the proper way for compile-time endianness-detection is. But looking around the codebase `_ZFS_LITTLE_ENDIAN` seems like the way to go. @pzakha ?

problame commented:
Actually I just noticed that `ztest -o` is broken as well.

behlendorf commented on lib/libzpool/util.c:
```diff
@@ -159,7 +159,7 @@ set_global_var(char *arg)
 	u_longlong_t val;
 
-#ifndef _LITTLE_ENDIAN
+#ifndef _ZFS_LITTLE_ENDIAN
```
It looks like `zdb_dump_block`  in `zdb.c` needs to be updated as well.  The `_ZFS` prefix was added by commit 5678d3f5 to resolve a conflict, it looks like we missed a couple instances.

problame commented:
@behlendorf I pushed an additional fix for ztest to this branch (unrelated to endianness). They should be committed separately.

I don't feel qualified to fix the endianness `#define`'s in the entire code base.

IMHO there should be a gobal constant / static inline function in zfs_context.h or similar that can be called to determine endianness. Compiler inlining and constant folding will do the rest.

problame commented:
Requesting review from @pzakha

behlendorf commented on cmd/ztest/ztest.c:
```diff
@@ -918,8 +924,21 @@ process_options(int argc, char **argv)
+			if (strlcpy(v, optarg, ZO_GVARS_MAX_ARGLEN) >=
+			    ZO_GVARS_MAX_ARGLEN) {
+				(void) fprintf(stderr,
+				    "global var option '%s' is tool long\n",
```
```suggestion
				    "global var option '%s' is too long\n",
```

behlendorf reviewed (commented):
This likely ended up broken because we don't have a test case which verifies this works correctly.  It'd be nice to add even just a very basic one.

nit: build failure on non-x86_84 platforms

```
ztest.c: In function ‘process_options’:
ztest.c:929:9: error: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘unsigned int’ [-Werror=format=]
         "max global var count (%lu) exceeded\n",
```

pzakha commented on cmd/ztest/ztest.c:
```diff
@@ -918,8 +924,21 @@ process_options(int argc, char **argv)
 			break;
 		case 'o':
-			if (set_global_var(optarg) != 0)
+			if (zo->zo_gvars_count > ZO_GVARS_MAX_COUNT) {
```
This should be `>=`

pzakha reviewed (approved):
Thanks for fixing this!

problame commented:
I addressed all the review comments in the last force-push.

>  This likely ended up broken because we don't have a test case which verifies this works correctly. It'd be nice to add even just a very basic one.

I think the ztest feature never worked for the child processes. Integration-testing `-o` for ztest would be tricky, I sadly can't afford the half-day to get that right :/

> 
> I don't feel qualified to fix the endianness #define's in the entire code base.
> 
> IMHO there should be a gobal constant / static inline function in zfs_context.h or similar that can be called to determine endianness. Compiler inlining and constant folding will do the rest.

What do we do about this?

behlendorf commented:
> IMHO there should be a gobal constant / static inline function in zfs_context.h or similar that can be called to determine endianness.

I'm not sure I follow, this is what the `_ZFS_LITTLE_ENDIAN` and `_ZFS_BIG_ENDIAN` defines are for.  They're defined in the `sys/isa_defs.h` file.  What am I missing?

problame commented:
> I'm not sure I follow, this is what the _ZFS_LITTLE_ENDIAN and _ZFS_BIG_ENDIAN defines are for. They're defined in the sys/isa_defs.h file. What am I missing?

I think it's a good idea to avoid conditional compilation inside functions wherever possible (at least the Linux kernel style guide agreees  https://www.kernel.org/doc/html/v4.10/process/coding-style.html#conditional-compilation ).

=> add a `static inline boolean_t is_little_endian() { ... }` in `sys/isa_defs.h` and use that in `set_global_var`.

Regardless, I think that'd be another PR, most likely not by me. I'll create an issue that references this comment.

problame commented:
Haha, there is an additional catch to this: `ztest` should forward the `-o` flags to `zdb`. I'll implement this and push it to this branch over the day.

problame commented:
Pushed:

```
commit 1454d5994428b35ee05aa8cd6a690a1fa3bf671a (HEAD -> zil-pmem-tmp, problame/fix-zdb--o)
Author: Christian Schwarz <me@cschwarz.com>
Date:   Thu Feb 18 12:20:09 2021 +0100

    ztest: propagate -o to the zdb child process
    
    I think this is the behavior that most users expect.
    
    Future work: have a separate flag, e.g., -O, to specify separate
    set_global_vars for the zdb child than for the ztest children.
    
    Signed-off-by: Christian Schwarz <me@cschwarz.com>

```
